File: app/flappybird/FlappyBird.py
import pygame
import random
import sys
import numpy as np

# Game constants
SCREEN_WIDTH = 1200
SCREEN_HEIGHT = 600
BIRD_RADIUS = 20
PIPE_WIDTH = 35
PIPE_GAP = 150
PIPE_VELOCITY = 3
GRAVITY = 0.4
INPUT_DIM = 3

# Game states
IDLE = 0
PLAYING = 1
GAME_OVER = 2


# Bird class
class Bird:
    def __init__(self, weights=None):
        self.x = SCREEN_WIDTH // 2
        self.y = SCREEN_HEIGHT // 2
        self.velocity = 0
        self.alive = True
        self.network = NeuralNetwork(weights)

    def flap(self):
        self.velocity = -7

    def update(self, inputs):
        if self.alive:
            output = self.network.feed_forward(inputs)
            if output > 0.5:
                self.flap()

        self.velocity += GRAVITY
        self.y += self.velocity

    def draw(self, screen):
        pygame.draw.circle(screen, (255, 255, 0), (self.x, int(self.y)), BIRD_RADIUS)

    def check_collision(self, pipes):
        if self.y - BIRD_RADIUS < 0 or self.y + BIRD_RADIUS > SCREEN_HEIGHT:
            return True

        for pipe in pipes:
            if pipe.collides_with_bird(self):
                return True

        return False


# Pipe class
class Pipe:
    def __init__(self, x):
        self.x = x
        self.height = random.randint(100, SCREEN_HEIGHT - PIPE_GAP - 100)

    def move(self):
        self.x -= PIPE_VELOCITY

    def draw(self, screen):
        pygame.draw.rect(screen, (0, 255, 0), (self.x, 0, PIPE_WIDTH, self.height))
        pygame.draw.rect(screen, (0, 255, 0), (self.x, self.height + PIPE_GAP, PIPE_WIDTH,
                                               SCREEN_HEIGHT - self.height - PIPE_GAP))

    def collides_with_bird(self, bird):
        if bird.y - BIRD_RADIUS < self.height or bird.y + BIRD_RADIUS > self.height + PIPE_GAP:
            if bird.x + BIRD_RADIUS > self.x and bird.x - BIRD_RADIUS < self.x + PIPE_WIDTH:
                return True
        return False


# Neural Network class
class NeuralNetwork:
    def __init__(self, weights=None):
        self.weights = weights if weights is not None else [np.random.randn(INPUT_DIM, 4), np.random.randn(4, 1)]

    def feed_forward(self, inputs):
        x = np.array(inputs).reshape(INPUT_DIM, 1)
        for weight in self.weights:
            x = NeuralNetwork.sigmoid(np.dot(weight.T, x))
        return x.item()

    def sigmoid(x):
        # Clamp the input values to prevent overflow
        x = np.clip(x, -500, 500)
        return 1 / (1 + np.exp(-x))
# Helper function for the sigmoid activation
# def sigmoid(x):
#     return 1 / (1 + np.exp(-x))


# Game class
# Game class
class FlappyBirdGame:
    def __init__(self, num_birds):
        self.state = IDLE
        self.epochs = 30
        self.current_epoch = 0
        self.num_birds = num_birds
        self.birds = []
        self.pipes = []
        self.pipe_count = 0
        self.high_score = 0
        self.cycle_count = 0
        self.best_weights = None

        # Pygame variables for the bird count display
        self.font = pygame.font.Font(None, 24)
        self.bird_count_text = self.font.render("", True, (255, 255, 255))
        self.bird_count_rect = self.bird_count_text.get_rect(topright=(10, 10))
        self.iter_gen_text = self.font.render("", True, (255, 255, 255))
        self.iter_gen_rect = self.iter_gen_text.get_rect(topright=(10, 40))
        self.high_score_text = self.font.render("", True, (255, 255, 255))
        self.high_score_rect = self.high_score_text.get_rect(topright=(10, 70))
        self.current_score_text = self.font.render("", True, (255, 255, 255))
        self.current_score_rect = self.current_score_text.get_rect(topright=(10, 100))

    def start_game(self):
        self.state = PLAYING
        self.current_epoch += 1
        self.birds = [Bird(weights=self.add_noise_to_weights(self.best_weights)) for _ in range(self.num_birds)]
        self.pipes = [Pipe(SCREEN_WIDTH)]
        self.cycle_count = 0
        global PIPE_VELOCITY
        #PIPE_VELOCITY *= 1.1
        global PIPE_GAP
        #PIPE_GAP = int(PIPE_GAP*0.95)

    def reset_game(self):
        self.state = IDLE
        self.pipes = []

    def update(self):
        if self.state == PLAYING:
            live_birds = 0
            for bird in self.birds:
                next_pipe = self.get_next_pipe()
                if next_pipe:
                    inputs = [next_pipe.x - bird.x, next_pipe.height + PIPE_GAP - bird.y, PIPE_VELOCITY]
                else:
                    # If there are no pipes, set inputs to a large value
                    inputs = [SCREEN_WIDTH, SCREEN_HEIGHT, PIPE_VELOCITY]

                bird.update(inputs)

                if bird.check_collision(self.pipes):
                    bird.alive = False

                if bird.alive:
                    live_birds += 1

            self.cycle_count += 1

            if live_birds == 1:
                self.update_best_weights()
            
            if live_birds == 0:
                self.state = GAME_OVER
                #reset pipe count
                self.pipe_count = 0

            if self.pipes[-1].x < SCREEN_WIDTH - 400:
                self.pipes.append(Pipe(SCREEN_WIDTH))
                self.pipe_count += 1
                #check if high score
                if self.pipe_count > self.high_score:
                    self.high_score = self.pipe_count

            for pipe in self.pipes:
                pipe.move()

            # Update bird count text
            self.bird_count_text = self.font.render(f"Birds Alive: {live_birds}", True, (255, 255, 255))
            self.iter_gen_text   = self.font.render(f"Generation: {self.current_epoch}", True, (255, 255, 255))
            self.high_score_text = self.font.render(f"High Score: {self.high_score}", True, (255, 255, 255))
            self.current_score_text = self.font.render(f"Current Score: {self.pipe_count}", True, (255, 255, 255))

    def get_next_pipe(self):
        for pipe in self.pipes:
            if pipe.x + PIPE_WIDTH > SCREEN_WIDTH // 2:
                return pipe
        return None

    def draw(self, screen):
        screen.fill((0, 0, 0))

        for pipe in self.pipes:
            pipe.draw(screen)

        for bird in self.birds:
            if bird.alive:
                bird.draw(screen)

        if self.state == GAME_OVER:
            font = pygame.font.Font(None, 36)
            text = font.render("Game Over", True, (255, 255, 255))
            text_rect = text.get_rect(center=(SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2))
            screen.blit(text, text_rect)

        # Draw bird count
        screen.blit(self.bird_count_text, self.bird_count_rect)
        screen.blit(self.iter_gen_text,   self.iter_gen_rect)
        screen.blit(self.high_score_text, self.high_score_rect)
        screen.blit(self.current_score_text, self.current_score_rect)

    def update_best_weights(self):
        for bird in self.birds:
            if bird.alive:
                self.best_weights = bird.network.weights
                break
    def add_noise_to_weights(self, weights):
        if weights is None:
            return None
        noise = [np.random.normal(0, 0.1, size=weight.shape) for weight in weights]
        return [weight + noise[i] for i, weight in enumerate(weights)]

# Game function
def play_game(num_birds):
    pygame.init()
    screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
    clock = pygame.time.Clock()

    game = FlappyBirdGame(num_birds)

    while game.current_epoch < game.epochs:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()

        game.update()
        game.draw(screen)

        pygame.display.update()
        clock.tick(5000)

        if game.state == IDLE:
            game.start_game()

        elif game.state == GAME_OVER:
            game.reset_game()

# Start the game
play_game(50)

File: verif/big_core/tests/FlappyBird.c
#include "../../../app/defines/big_core_defines.h"
#include "../../../app/defines/graphic_vga.h"
// Game constants
#define BIRD_RADIUS 4
#define PIPE_WIDTH 10
#define PIPE_GAP 15
#define PIPE_VELOCITY 3
#define GRAVITY 0
#define INPUT_DIM 3
#define NULL ((void*)0)

// Game states
#define IDLE 0
#define PLAYING 1
#define GAME_OVER 2

// Neural Network struct
typedef struct {
    int weights[INPUT_DIM][3];
} NeuralNetwork;


// Bird struct
typedef struct {
    int x;
    int y;
    int velocity;
    int alive;
    NeuralNetwork network; // Add the neural network weights here
} Bird;

// Pipe struct
typedef struct {
    int x;
    int height;
} Pipe;

// Flappy Bird Game struct
typedef struct {
    int state;
    int epochs;
    int currentEpoch;
    Bird birds[49]; // Assuming a maximum of 100 birds for demonstration purposes
    Pipe pipes[99]; // Assuming a maximum of 100 pipes for demonstration purposes
    int numBirds;
    int numPipes;
    unsigned int * seed;
    NeuralNetwork bestWeights; // Pointer to the best weights
} FlappyBirdGame;

// Helper function declarations
int sigmoid(int x);
int clamp(int value, int min, int max);
double exp(double x);
unsigned int myRand(unsigned int* seed);

// Bird function declarations
void flap(Bird *bird);
void updateBird(Bird *bird, int *inputs);
void drawBird(Bird *bird);
int checkCollision(Bird *bird, Pipe *pipes, int numPipes);

// Pipe function declarations
void movePipe(Pipe *pipe);
void drawPipe(Pipe *pipe);
int collidesWithBird(Pipe *pipe, Bird *bird);

// Neural Network function declarations
int feedForward(NeuralNetwork *network, int *inputs);

// Game function declarations
void startGame(FlappyBirdGame * game);
void resetGame(FlappyBirdGame * game);
void updateGame(FlappyBirdGame * game);
void drawGame(FlappyBirdGame * game);
Pipe* getNextPipe(FlappyBirdGame * game);
void updateBestWeights(FlappyBirdGame * game);
void addNoiseToWeights(NeuralNetwork *weights, unsigned int *seed);


// Bird function definitions
void flap(Bird *bird) {
    bird->velocity = -7;
}

void updateBird(Bird *bird, int *inputs) {
    if (bird->alive) {
        flap(bird);
       // int output = feedForward(&bird->network, inputs);
       // if (output > 0.5) {
       //     flap(bird);
       // }
    }

    bird->velocity += GRAVITY;
    bird->y += bird->velocity;
}

int checkCollision(Bird *bird, Pipe *pipes, int numPipes) {
    if (bird->y - BIRD_RADIUS < 0 || bird->y + BIRD_RADIUS > SCREEN_HEIGHT) {
        return 1;
    }

    for (int i = 0; i < numPipes; i++) {
        if (collidesWithBird(&pipes[i], bird)) {
            return 1;
        }
    }

    return 0;
}

void drawBird(Bird *bird) {
    // Draw bird circle
    // Ignoring color for now
    // Use bird->x, bird->y, BIRD_RADIUS for drawing
    draw_circle(bird->x, bird->y, BIRD_RADIUS, 1);
}


// Pipe function definitions
void movePipe(Pipe *pipe) {
    pipe->x -= PIPE_VELOCITY;
}

int collidesWithBird(Pipe *pipe, Bird *bird) {
    if (bird->y - BIRD_RADIUS < pipe->height || bird->y + BIRD_RADIUS > pipe->height + PIPE_GAP) {
        if (bird->x + BIRD_RADIUS > pipe->x && bird->x - BIRD_RADIUS < pipe->x + PIPE_WIDTH) {
            return 1;
        }
    }
    return 0;
}

int sigmoid(int x) {
    if (x < -500) {
        return 0;
    }
    else if (x > 500) {
        return 1;
    }
    else {
        int exp_val = exp(x);
        return exp_val / (exp_val + 1);
    }
}

int clamp(int value, int min, int max) {
    if (value < min) {
        return min;
    }
    else if (value > max) {
        return max;
    }
    else {
        return value;
    }
}

double exp(double x) {
    double result = 1.0;
    double term = 1.0;
    int i;

    for (i = 1; i <= 10; ++i) {
        term *= x / i;
        result += term;
    }

    return result;
}

void drawPipe(Pipe *pipe) {
    // Draw pipe rectangles
    // Ignoring color for now
    // Use pipe->x, pipe->height, PIPE_WIDTH, PIPE_GAP, SCREEN_HEIGHT for drawing

    // Upper pipe rectangle
    Rectangle upperRect = {pipe->x, 0, PIPE_WIDTH, pipe->height};
    draw_rectangle(&upperRect, 1);

    // Lower pipe rectangle
    Rectangle lowerRect = {pipe->x, pipe->height + PIPE_GAP, PIPE_WIDTH, SCREEN_HEIGHT - pipe->height - PIPE_GAP};
    draw_rectangle(&lowerRect, 1);
}

// Neural Network function definitions
int feedForward(NeuralNetwork *network, int *inputs) {
    int layer1[4];
    int output;

    // Perform feed-forward calculations
    for (int i = 0; i < 4; i++) {
        int sum = 0;
        for (int j = 0; j < INPUT_DIM; j++) {
            sum += inputs[j] * network->weights[j][i];
        }
        layer1[i] = sigmoid(sum);
    }

    int sum = 0;
    for (int i = 0; i < 4; i++) {
        sum += layer1[i] * network->weights[1][i];
    }
    output = sigmoid(sum);

    return output;
}

// Game function definitions
void startGame(FlappyBirdGame* game) {
    rvc_printf("GAME\n");
    game->state = PLAYING;
    game->currentEpoch += 1;

    for (int i = 0; i < game->numBirds; i++) {
        // Initialize birds with weights and other properties
        game->birds[i].x = SCREEN_WIDTH / 2;
        game->birds[i].y = SCREEN_HEIGHT / 2;
        game->birds[i].velocity = 0;
        game->birds[i].alive = 1;
        // Initialize the neural network weights for each bird
        // Use the best weights or add noise as needed
    }

    game->pipes[0].x = SCREEN_WIDTH;
    // Generate a random height for the first pipe
    game->pipes[0].height = myRand(game->seed) % (SCREEN_HEIGHT - PIPE_GAP - 100) + 100;
    game->numPipes = 1;

}

void resetGame(FlappyBirdGame* game) {
    game->state = IDLE;
    game->numPipes = 0;
}

void updateGame(FlappyBirdGame* game) {

    rvc_printf("UPDATE\n");
    rvc_print_int(game->state);
    rvc_printf("\n");
    if (game->state == PLAYING) {
        int liveBirds = 0;
        rvc_printf("IN IF1\n");
        for (int i = 0; i < game->numBirds; i++) {
            //rvc_print_int(i);
            //rvc_printf(" ");
            Pipe *nextPipe = getNextPipe(game);
            int inputs[INPUT_DIM];

            if (nextPipe) {
                inputs[0] = nextPipe->x - game->birds[i].x;
                inputs[1] = nextPipe->height + PIPE_GAP - game->birds[i].y;
                inputs[2] = PIPE_VELOCITY;
            }
            else {
                inputs[0] = SCREEN_WIDTH;
                inputs[1] = SCREEN_HEIGHT;
                inputs[2] = PIPE_VELOCITY;
            }
            // rvc_printf("BEFORE UPDATE BIRD");

            updateBird(&game->birds[i], inputs);
            // rvc_printf("AFTER UPDATE BIRD");

            if (checkCollision(&game->birds[i], game->pipes, game->numPipes)) {
                game->birds[i].alive = 0;
            }

            if (game->birds[i].alive) {
                liveBirds++;
            }
        }

        if (liveBirds == 1) {
            updateBestWeights(game);
        }

        if (liveBirds == 0) {
            game->state = GAME_OVER;
        }

        if (game->pipes[game->numPipes - 1].x < SCREEN_WIDTH - 400) {
            game->pipes[game->numPipes].x = SCREEN_WIDTH;
            game->pipes[game->numPipes].height = myRand(game->seed) % (SCREEN_HEIGHT - PIPE_GAP - 100) + 100;
            game->numPipes++;
        }
    }
}

void drawGame(FlappyBirdGame* game) {
    clear_screen();
    for(int i=0 ; i<game->numBirds; i++) {
        drawBird(&game->birds[i]);
    }
    for(int i=0 ; i<game->numPipes; i++) {
        drawPipe(&game->pipes[i]);
    }
}

Pipe* getNextPipe(FlappyBirdGame* game) {
    for (int i = 0; i < game->numPipes; i++) {
        if (game->pipes[i].x + PIPE_WIDTH > SCREEN_WIDTH / 2) {
            return &game->pipes[i];
        }
    }
    return NULL;
}

void updateBestWeights(FlappyBirdGame* game) {
    for (int i = 0; i < game->numBirds; i++) {
        if (game->birds[i].alive) {
            // Update the best weights using the bird's network weights
            // Store the best weights for future use
            game->bestWeights = game->birds[i].network;
            break;
        }
    }
}

unsigned int myRand(unsigned int* seed) {
    *seed = (*seed * 1103515245 + 12345) & 0x7FFFFFFF;
    return *seed;
}

void addNoiseToWeights(NeuralNetwork *network, unsigned int *seed) {
    // Define the noise level (scaled by 1000 to simulate floating-point precision)
    int noiseLevel = 300; // Equivalent to 0.3

    // Iterate over the weights array and add random noise to each weight
    for (int i = 0; i < INPUT_DIM; i++) {
        for (int j = 0; j < 4; j++) {
            int randomValue = myRand(seed);
            double noise = (randomValue % (2 * noiseLevel + 1) - noiseLevel) / 1000.0;
            network->weights[i][j] += noise;
        }
    }
}

// Main function
int main() {
    clear_screen();
    rvc_printf("STARTED\n");
    FlappyBirdGame game;
    game.numBirds = 50;
    game.epochs = 20;
    game.currentEpoch = 0;
    rvc_print_int(game.epochs);
    rvc_printf("\n");
    rvc_print_int(game.currentEpoch);
    rvc_printf("\n");
    *(game.seed) = 42;

    startGame(&game);
    rvc_printf("RETURNED\n");

    while (game.currentEpoch < game.epochs) {
        rvc_printf("WHILE\n");
        
        updateGame(&game);
        rvc_printf("AFTER UPDATE\n");
    
        drawGame(&game);

        if (game.state == IDLE) {
            startGame(&game);
        }
        else if (game.state == GAME_OVER) {
            resetGame(&game);
        }
    }

    return 0;
}
